lib/r10k/tarball.rb

```diff
@@ -0,0 +1,257 @@
+      @source = source
+      @sha256digest = sha256digest
+    end
+
+    # @return [String] Directory. Where the cache will be created.
+    def cache_root
+      File.join(settings[:cache_root], sanitized_dirname(source))
+    end
+
+    # @return [String] File. The path the tarball will be cached to.
+    def cache_path
+      File.join(cache_root, tarball_basename)
+    end
+
+    # @return [String] The basename of the tarball cache file.
+    def tarball_basename
+      name + '.tar.gz'
+    end
+
+    # Extract the cached tarball to the target directory.
+    #
+    # @param target_dir [String] Where to unpack the tarball
+    def unpack(target_dir)
+      file = File.open(cache_path, 'rb')
+      reader = Zlib::GzipReader.new(file)
+      begin
+        Minitar.unpack(reader, target_dir)
+      ensure
+        reader.close
+      end
+    end
+
+    # @param target_dir [String] The directory to check if is insync with the
+    #        tarball content
+    #
+    # @return [Boolean]
+    def insync?(target_dir, purge: true)
+      target_dir_entries = Find.find(target_dir).map(&:to_s) - [target_dir]
+      each_tarball_entry do |entry|
+        found = target_dir_entries.delete(File.join(target_dir, entry.full_name.chomp('/')))
+        return false if found.nil?
+        next if entry.directory?
+        return false unless file_digest(found) == reader_digest(entry)
+      end
+      purge ? target_dir_entries.empty? : true
+    end
+
+    # Download the tarball from @source to @cache_path
+    def download
+      Dir.mktmpdir do
+        tempfile = Tempfile.new(tarball_basename)
+        tempfile.binmode
+        src_uri = URI.parse(source)
+
+        temp_digest = case src_uri.scheme
+                      when 'file', nil
+                        copy_to_file(src_uri.path, tempfile)
+                      when %r{[a-z]} # Windows drive letter
+                        copy_to_file(src_uri.to_s, tempfile)
+                      when %r{https?}
+                        download_to_file(src_uri, tempfile)
+                      else
+                        raise "Unexpected source scheme #{src_uri.scheme}"
+                      end
+
+        # Verify the download
+        unless (sha256digest == temp_digest) || sha256digest.nil?
+          raise 'Downloaded file does not match digest'
+        end
+
+        # Move the download to cache_path
+        FileUtils::mkdir_p(cache_root)
+        begin
+          FileUtils.mv(tempfile.path, cache_path)
+        rescue Errno::EACCES
+          FileUtils.cp(tempfile.path, cache_path)
+        end
+
+        tempfile.close
+        tempfile.unlink
+      end
+    end
+
+    # Is the cached tarball valid? Checks the cached tarball's digest against
+    # the expected digest.
+    #
+    # @return [Boolean]
+    def cache_valid?
+      unless sha256digest.nil?
+        File.exist?(cache_path) && sha256digest == file_digest(cache_path)
+      else
+        File.exist?(cache_path)
+      end
```

Comment:
I would switch these around and use an `if` instead of `unless`. Using `unless` with `else` is hard to read imo.

Reply:
or you could just do
```suggestion
      return false unless File.exist?(cache_path)
      return false if sha256digest && sha256digest != file_digest(cache_path)
      return true
```

Reply:
Good call. Adding a commit to simplify.